### PR TITLE
ES6 Rewrite and Unit Tests

### PR DESCRIPTION
**Pull request by KanoYugoro**
Updated to use ES6 syntax, as well as get off of grub and change to mocha and chai. The tests are not ideal as they do not entirely mock the expected responses, but are approximately equivalent to what existed previously. They are written in ES6 javascript as well now, instead of coffeescript.

All functions requiring callbacks were rewritten using promises, and massive code duplication was reduced wherever found.

A strict eslint version was added to the repository to help enforce coding consistency, as it was sort of all over the place. It reads easier now.

The build agent will fail hard when it tries to build this, as the previously existing version used an entirely different method to run unit tests. Configuring this properly is the last remaining task for this branch.

### CODE REVIEW

**Comment by mtscout6** [comment]
In the `.eslintignore` file can you also ignore the lib folder since the compiled babel output would go there?

File: .travis.yml [hunk 1]
```diff
@@ -1,5 +1,4 @@
 language: node_js
 node_js:
-   - "0.10"
-before_install: npm install -g grunt-cli
+   - "4.2"
 install: npm install
```
Comment (mtscout6) [inline .travis.yml]: `npm install` is done by default on travis, this line isn't needed. Consider our commit hook repo as a template here: https://github.com/TWExchangeSolutions/jira-precommit-hook/blob/master/.travis.yml

In particular the `sudo: false` which will ensure the builds run faster and are queued for a shorter period of time.

File: src/jira.js [hunk 2]
```diff
@@ -0,0 +1,1016 @@
+import request from 'request-promise';
+import url from 'url';
+
+/**
+ * Wrapper for the JIRA Rest Api
+ * @constructor
+ * @param {object} options - All other parameters to this function are fields on this object.
+ * @param {string} protocol - What protocol to use to connect to jira? Ex: http|https
```
Comment (mtscout6) [inline src/jira.js]: See: http://usejsdoc.org/tags-param.html#parameters-with-properties

File: src/jira.js [hunk 3]
```diff
@@ -0,0 +1,1016 @@
...
+  // maxResults: The maximum number of users to return (defaults to 50).
+  // includeActive: If true, then active users are included in the results (default true)
+  // includeInactive: If true, then inactive users are included in the results (default false)
+  // ### Returns ###
+  // *  users: array of users for the user
+  // [Jira Doc](http://docs.atlassian.com/jira/REST/latest/#d2e3756)
+  searchUsers({username, startAt, maxResults, includeActive, includeInactive}) {
+    const requestOptions = {
+      rejectUnauthorized: this.strictSSL,
+      uri: this.makeUri(`/user/search?username=${username}&startAt=${startAt || 0}&maxResults=${maxResults || 50}&includeActive=${includeActive || true}&includeInactive=${includeInactive || false}`),
```
Comment (mtscout6) [inline src/jira.js]: This url can be cleaned up with the querystring serialization options. See: https://www.npmjs.com/package/request#request-options-callback

File: src/jira.js [hunk 4]
```diff
@@ -0,0 +1,1016 @@
...
+   *     "realAssigneeType": "PROJECT_DEFAULT",
+   *     "realAssignee": {
+   *         "self": "http://localhostname:8090/jira/rest/api/2.0/user?username=user@domain.com",
+   *         "name": "user@domain.com",
+   *         "displayName": "User name",
+   *         "active": true
+   *     },
+   *     "isAssigneeTypeValid": true
+   * }]
+   */
```
Comment (mtscout6) [inline src/jira.js]: You started using JSDocs, but this isn't. Can that be more consistent?

File: test/jira-tests.js [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+      const jira = new JiraApi({
+        protocol: 'http',
+        host: 'jira.somehost.com',
+        port: '8080',
+        username: 'someusername',
+        password: 'somepassword',
+        apiVersion: '2.0',
+        strictSSL: true,
+        request: dummyRequest
+      });
```
Comment (mtscout6) [inline test/jira-tests.js]: You seem to be using a lot of the same default options in your tests. Have you considered using a default options variable, and override them as needed?

File: src/jira.js [hunk 6]
```diff
@@ -0,0 +1,714 @@
...
+
+  /** Get a list of Sprints belonging to a Rapid View
+   *  @param {string} rapidViewId - the id for the rapid view
+   */
+  getLastSprintForRapidView(rapidViewId) {
+    const requestOptions = {
+      rejectUnauthorized: this.strictSSL,
+      uri: this.makeUri(`/sprintquery/${rapidViewId}`),
+      method: 'GET',
+      json:true
```
Comment (TrevorMartz) [inline src/jira.js]: There should be a space before true.

File: src/jira.js [hunk 7]
```diff
@@ -0,0 +1,714 @@
...
+   *  @param {string} issueId - the id of the existing issue
+   *  @param {string} sprintId - the id of the sprint to add it to
+   */
+  addIssueToSprint(issueId, sprintId) {
+    const requestOptions = {
+      rejectUnauthorized: this.strictSSL,
+      uri: this.makeUri(`/sprint/${sprintId}/issues/add`),
+      method: 'PUT',
+      followAllRedirects: true,
+      json:true,
```
Comment (TrevorMartz) [inline src/jira.js]: Space before true here too.

File: src/jira.js [hunk 8]
```diff
@@ -0,0 +1,714 @@
...
+      uri: this.makeUri('/issuetype'),
+      method: 'GET',
+      json: true
+    };
+
+    return this.doRequest(requestOptions);
+  }
+
+  /** Register a webhook
+   *  webhook: properly formatted webhook
```
Comment (TrevorMartz) [inline src/jira.js]: Missing the @param here.

File: src/jira.js [hunk 9]
```diff
@@ -0,0 +1,714 @@
...
+      uri: this.makeUri(`/webhook/${webhookID}`),
+      method: 'GET',
+      json: true
+    };
+
+    return this.doRequest(requestOptions);
+  }
+
+  /** Delete a registered webhook
+   *  webhookID: id of the webhook to delete
```
Comment (TrevorMartz) [inline src/jira.js]: Missing the @param here as well.

File: src/jira.js [hunk 10]
```diff
@@ -0,0 +1,714 @@
...
+      uri: this.makeUri('/session'),
+      method: 'GET',
+      json: true
+    };
+
+    return this.doRequest(requestOptions);
+  }
+
+  /** Retrieve the backlog of a certain Rapid View
+   *  rapidViewId: rapid view id
```
Comment (TrevorMartz) [inline src/jira.js]: And another @param.

File: src/jira.js [hunk 11]
```diff
@@ -0,0 +1,801 @@
...
+      }
+    };
+
+    return this.doRequest(requestOptions);
+  }
+
+  /** Search user on Jira
+   *  @name searchUsers
+   *  @function
+   * @param {string} username - A query string used to search username, name or e-mail address
```
Comment (TrevorMartz) [inline src/jira.js]: Inconsistent spacing between the * and @ here and several other places.

**Comment by mtscout6** [comment]
:+1: Good cleanup

**Pull request merged**